File: src/CodexiumMagnus.Core/Reporting/ReportWriter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;

namespace CodexiumMagnus.Core.Reporting
{
    /// <summary>
    /// Central writer for all reports so build, test, and runtime tools look the same.
    /// </summary>
    public sealed class ReportWriter
    {
        private readonly List<ReportEntry> _entries = new();

        public IReadOnlyList<ReportEntry> Entries => _entries;

        public void Add(ReportEntry entry)
        {
            if (entry is null) throw new ArgumentNullException(nameof(entry));
            _entries.Add(entry);
        }

        public void Add(ReportSeverity severity, string title, string? details = null, string? source = null)
        {
            _entries.Add(new ReportEntry
            {
                Severity = severity,
                Title = title,
                Details = details,
                Source = source
            });
        }

        public void WriteMarkdown(string path)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);

            var sb = new StringBuilder();
            sb.AppendLine("# Codexium Magnus Report");
            sb.AppendLine();
            sb.AppendLine($"Generated: {DateTime.UtcNow:O}");
            sb.AppendLine();
            sb.AppendLine("| Time (UTC) | Severity | Source | Title |");
            sb.AppendLine("|------------|----------|--------|-------|");

            foreach (var e in _entries)
            {
                sb.Append("| ")
                  .Append(e.TimestampUtc.ToString("O"))
                  .Append(" | ")
                  .Append(e.Severity)
                  .Append(" | ")
                  .Append(string.IsNullOrWhiteSpace(e.Source) ? "-" : e.Source)
                  .Append(" | ")
                  .Append(e.Title.Replace("|", "/"))
                  .AppendLine(" |");
            }

            File.WriteAllText(path, sb.ToString(), Encoding.UTF8);
        }

        public void WriteJson(string path)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);

            var json = JsonSerializer.Serialize(_entries, new JsonSerializerOptions
            {
                WriteIndented = true
            });

            File.WriteAllText(path, json, Encoding.UTF8);
        }
    }
}

File: src/codexium-magnus-core/Reporting/ReportWriter.cpp
#include "ReportWriter.h"
#include <QFile>
#include <QDir>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QTextStream>
#include <QDateTime>

namespace CodexiumMagnus::Core::Reporting {

void ReportWriter::add(const ReportEntry& entry) {
    m_entries.append(entry);
}

void ReportWriter::add(ReportSeverity severity, const QString& title, 
                       const QString& details, const QString& source) {
    ReportEntry entry;
    entry.severity = severity;
    entry.title = title;
    entry.details = details;
    entry.source = source;
    entry.timestampUtc = QDateTime::currentDateTimeUtc();
    m_entries.append(entry);
}

void ReportWriter::writeMarkdown(const QString& path) const {
    QFileInfo fileInfo(path);
    QDir dir = fileInfo.absoluteDir();
    if (!dir.exists()) {
        dir.mkpath(".");
    }

    QFile file(path);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        return;
    }

    QTextStream out(&file);
    out << "# Codexium Magnus Report\n\n";
    out << "Generated: " << QDateTime::currentDateTimeUtc().toString(Qt::ISODate) << "\n\n";
    out << "| Time (UTC) | Severity | Source | Title |\n";
    out << "|------------|----------|--------|-------|\n";

    for (const auto& entry : m_entries) {
        QString severityStr;
        switch (entry.severity) {
            case ReportSeverity::Info:
                severityStr = "Info";
                break;
            case ReportSeverity::Warning:
                severityStr = "Warning";
                break;
            case ReportSeverity::Fatal:
                severityStr = "Fatal";
                break;
        }

        QString sourceStr = entry.source.isEmpty() ? "-" : entry.source;
        QString titleStr = entry.title;
        titleStr.replace("|", "/");

        out << "| " << entry.timestampUtc.toString(Qt::ISODate)
            << " | " << severityStr
            << " | " << sourceStr
            << " | " << titleStr
            << " |\n";
    }
}

void ReportWriter::writeJson(const QString& path) const {
    QFileInfo fileInfo(path);
    QDir dir = fileInfo.absoluteDir();
    if (!dir.exists()) {
        dir.mkpath(".");
    }

    QJsonArray entriesArray;
    for (const auto& entry : m_entries) {
        QJsonObject entryObj;
        entryObj["timestampUtc"] = entry.timestampUtc.toString(Qt::ISODate);
        
        QString severityStr;
        switch (entry.severity) {
            case ReportSeverity::Info:
                severityStr = "Info";
                break;
            case ReportSeverity::Warning:
                severityStr = "Warning";
                break;
            case ReportSeverity::Fatal:
                severityStr = "Fatal";
                break;
        }
        entryObj["severity"] = severityStr;
        entryObj["title"] = entry.title;
        entryObj["details"] = entry.details;
        entryObj["source"] = entry.source;
        
        entriesArray.append(entryObj);
    }

    QJsonDocument doc(entriesArray);
    QFile file(path);
    if (file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        file.write(doc.toJson());
    }
}

} // namespace CodexiumMagnus::Core::Reporting